src/main/java/org/prebid/server/settings/CompositeApplicationSettings.java

```diff
@@ -0,0 +1,144 @@
+            final BiFunction<String, Timeout, Future<String>> nextRetriever =
+                    next != null ? next::getAdUnitConfigById : null;
+            return getConfig(adUnitConfigId, timeout, applicationSettings::getAdUnitConfigById, nextRetriever);
+        }
+
+        private static <T> Future<T> getConfig(String key, Timeout timeout,
+                                               BiFunction<String, Timeout, Future<T>> retriever,
+                                               BiFunction<String, Timeout, Future<T>> nextRetriever) {
+            return retriever.apply(key, timeout)
+                    .recover(throwable -> nextRetriever != null
+                            ? nextRetriever.apply(key, timeout)
+                            : Future.failedFuture(throwable));
+        }
+
+        @Override
+        public Future<StoredRequestResult> getStoredRequestsById(Set<String> ids, Timeout timeout) {
+            final BiFunction<Set<String>, Timeout, Future<StoredRequestResult>> nextRetriever =
+                    next != null ? next::getStoredRequestsById : null;
+            return getStoredRequests(ids, timeout, applicationSettings::getStoredRequestsById, nextRetriever);
+        }
+
+        @Override
+        public Future<StoredRequestResult> getStoredRequestsByAmpId(Set<String> ids, Timeout timeout) {
+            final BiFunction<Set<String>, Timeout, Future<StoredRequestResult>> nextRetriever =
+                    next != null ? next::getStoredRequestsByAmpId : null;
+            return getStoredRequests(ids, timeout, applicationSettings::getStoredRequestsByAmpId, nextRetriever);
+        }
+
+        private static Future<StoredRequestResult> getStoredRequests(
+                Set<String> ids, Timeout timeout,
+                BiFunction<Set<String>, Timeout, Future<StoredRequestResult>> retriever,
+                BiFunction<Set<String>, Timeout, Future<StoredRequestResult>> nextRetriever) {
+
+            return retriever.apply(ids, timeout)
+                    .compose(retrieverResult ->
+                            nextRetriever == null || retrieverResult.getStoredIdToJson().size() == ids.size()
+                                    ? Future.succeededFuture(retrieverResult)
+                                    : getRemainingStoredRequests(ids, timeout, retrieverResult.getStoredIdToJson(),
+                                    nextRetriever));
+        }
+
+        private static Future<StoredRequestResult> getRemainingStoredRequests(
+                Set<String> ids, Timeout timeout, Map<String, String> fetchedStoredIdToJson,
+                BiFunction<Set<String>, Timeout, Future<StoredRequestResult>> retriever) {
+
+            return retriever.apply(subtractSets(ids, fetchedStoredIdToJson.keySet()), timeout)
+                    .compose(result -> Future.succeededFuture(StoredRequestResult.of(
+                            combineMaps(fetchedStoredIdToJson, result.getStoredIdToJson()), result.getErrors())));
```

Comment:
What happens to the errors reported by intermediary backends in chain?

Reply:
We just don't carry on them. If we could not obtain all stored requests from one retriever then we'll try the next one if it exists. This next retriever will accept the list of IDs for which the previous  failed.
So, the last fetcher will give the result with list of error with all missing IDs and we don't need of any additional actions.